Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp

```diff
@@ -243,13 +372,71 @@ void KafkaEventStreamDecoder::eventDataFromMessage(const std::string &buffer) {
   } else {
     periodBuffer = m_localEvents[0];
   }
-  for (decltype(nEvents) i = 0; i < nEvents; ++i) {
-    auto &spectrum = periodBuffer->getSpectrum(
-        m_specToIdx[static_cast<int32_t>(detData[i])]);
-    spectrum.addEventQuickly(TofEvent(static_cast<double>(tofData[i]) *
-                                          1e-3, // nanoseconds to microseconds
-                                      pulseTime));
+
+  auto numGroups = tofData.size() / MAX_BUFFERGROUP_EVENTS;
+  std::vector<size_t> sizes(numGroups + 1, MAX_BUFFERGROUP_EVENTS);
+  sizes[numGroups] = tofData.size() % MAX_BUFFERGROUP_EVENTS;
+  size_t start = 0;
+
+  auto starttime = std::chrono::system_clock::now();
+  for (auto &grpSize : sizes) {
+    if (grpSize == 0)
+      continue;
+
+    BufferGroup buffGrp;
+    buffGrp.tof.resize(grpSize);
+    buffGrp.detID.resize(grpSize);
+
+    auto end = start + grpSize;
+    std::copy(tofData.data() + start, tofData.data() + end,
+              buffGrp.tof.begin());
+    std::copy(detData.data() + start, detData.data() + end,
+              buffGrp.detID.begin());
+
+    buffGrp.pulseTime = pulseTime;
+
+    buffGrp.periodBuffer = periodBuffer;
+    m_buffer.emplace_back(buffGrp);
+    start += grpSize;
+  }
+
+  auto endTime = std::chrono::system_clock::now();
+  std::chrono::duration<double> dur = endTime - starttime;
+  totalEventFromMessageDuration += dur.count();
+  numEventFromMessageCalls += 1;
+}
+
+void KafkaEventStreamDecoder::populateEventDataWorkspace() {
+  auto startTime = std::chrono::system_clock::now();
+  for (auto &grp : m_buffer) {
+    auto nEvents = grp.tof.size();
+
+    if (nEvents == 0)
+      continue;
+
+    auto specindices = getSpectrumIndices(grp.detID.data(), nEvents);
+    auto perm = getSortedSpectraPermutation(specindices);
+    auto ranges = getRanges(perm, specindices);
```

Comment:
Do I understand correctly that this is independently sorting each chunk (do I understand correctly that there is at least 1 chunk per message)? Isn't that losing most of the intended benefit of sorting?

Have you tried doing this in the way we did in the prototype/benchmark, which gave us a ~5x speedup, i.e.,:
- Put *all* events from *all* messages (up to a threshold, potentially a per-bank buffer, since those are sorted already) into *one big buffer*.
- Sort big buffer in parallel.
- Insert events in parallel (this will let you move the `PARALLEL_FOR` to the "outer" loop, instead of having it on the inner one, or rather, there might be only a single loop).